extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -246,26 +261,152 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
-      // 100vw, but is fixed here so that future resizes of the viewport don't
-      // affect it.
-      const viewportSize = this.getViewport().getSize();
-      return this.attemptChangeSize(
-        AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
-          this.autoFormat_,
-          viewportSize,
-          this.element
-        ),
-        viewportSize.width
-      ).catch(() => {});
+    let upgradePromise = Promise.resolve();
+    // Note: this must happen before this.autoFormat_ is populated
+    // since it may change the outcome.
+    if (this.isInAdSizeOptimizationExperimentBranch()) {
+      upgradePromise = this.maybeUpgradeToResponsive();
     }
-    // This should happen last, as some diversion criteria rely on some of the
-    // preceding logic (specifically responsive logic).
-    this.divertExperiments();
-    this.maybeAddSinglePassExperiment();
+
+    return upgradePromise.then(() => {
+      this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
+
+      if (this.isResponsive_()) {
+        // Attempt to resize to the correct height. The width should already be
+        // 100vw, but is fixed here so that future resizes of the viewport don't
+        // affect it.
+        const viewportSize = this.getViewport().getSize();
+        return this.attemptChangeSize(
+          AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+            this.autoFormat_,
+            viewportSize,
+            this.element
+          ),
+          viewportSize.width
+        ).catch(() => {});
+      }
+
+      // This should happen last, as some diversion criteria rely on some of the
+      // preceding logic (specifically responsive logic).
+      this.divertExperiments();
+      this.maybeAddSinglePassExperiment();
+    });
+  }
+
+  /**
+   * Upgrades the ad unit to responsive if there is an opt-in setting in localstorage.
+   * See https://github.com/ampproject/amphtml/issues/23568 for design.
+   * @return {Promise} a promise that resolves when any upgrade is complete.
+   */
+  maybeUpgradeToResponsive() {
+    // If the ad unit is already responsive, there's nothing to do.
+    // Note that we can't use this.autoFormat_ here because it's not
+    // populated yet.
+    if (this.element.hasAttribute('data-auto-format')) {
+      return Promise.resolve();
+    }
+    return (
+      Services.storageForDoc(this.element)
+        .then(storage => storage.get(this.getAdSizeOptimizationStorageKey()))
+        .then(isAdSizeOptimizationEnabled => {
+          if (isAdSizeOptimizationEnabled) {
+            this.upgradeToResponsive();
+          }
+        })
+        // Do nothing if we fail to read localstorage.
+        .catch(() => {
+          dev().warn(TAG, 'Failed to look up publisher ad size settings.');
+        })
+    );
+  }
+
+  /**
+   * Actually upgrades the ad unit to responsive.
+   */
+  upgradeToResponsive() {
+    this.element.setAttribute('height', ADSENSE_RSPV_WHITELISTED_HEIGHT);
```

Comment:
I'm a bit unsure. It seems at some point styles are set with css on the element which means these attributes don't actually matter for layout any more, but not entirely sure if that has happened yet in buildCallback?

We do need to change the values in the attributes, because otherwise the element validation (as it currently works) will fail (isValidElement will throw later on).

Reply:
OK please add @lannka  for review

Reply:
AMP layout should have happened before this. But could you pls double confirm locally? (for example print logs that measures the height before / after this call.

Reply:
Done! Added console.error at the top/bottom of method printing
`this.win.getComputedStyle(this.element, null).height`
and it's the same value for both calls.